### PR TITLE
Update README.

### PR DESCRIPTION
**Pull request by mschaef-da**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,59 +1,72 @@
...
+
+
+## Integration Packaging and Deployment
+
+Integrations are packaged in [DIT files](https://github.com/digital-asset/daml-dit-api),
+built using the [`ddit` build tool](https://github.com/digital-asset/daml-dit-ddit),
+and can be deployed into Daml Hub using the same upload mechanism as other
+artifacts (Daml Models, bots, etc.). Daml Hub also has an 'arcade' facility
+that uses a public [GitHub repository](https://github.com/digital-asset/daml-dit-arcade-index)
+to maintain list of integrations and sample apps that can be deployed through
```
Comment (da-tanabe) [inline README.md]: to maintain *a* list

File: README.md [hunk 2]
```diff
@@ -1,59 +1,72 @@
...
-`dabl-meta.yaml` that describes the properties of a given
-integration. The metadata section includes the name of the entry point
-function for the integration, some descriptive text, and a list of
-all of the configuration arguments that the integration accepts:
+The integration framework API has two parts. First, there is an
+additional `integration_types` section within `dabl-meta.yaml` that
+lists and describes the integration types defined within the DIT file.
+This metadata section includes the name of the entry point function
+for the integration type, some descriptive text, and a list of all of
+the configuration arguments that the integration accepts:
```
Comment (da-tanabe) [inline README.md]: ~For documentation, I prefer bulleted/numeric lists where possible to break up the visual monotony of walls of text:~

```md
The integration framework API has two parts:
1. An additional `integration_types` section within `dabl-meta.yaml`...
```

Where are the two parts? I only see one.

File: README.md [hunk 3]
```diff
@@ -1,59 +1,72 @@
...
 
-The ledger event log integration is defined like this:
+The [ledger event log integration](https://github.com/digital-asset/daml-dit-integration-core/blob/master/src/core_int/integration_ledger_event_log.py) is defined like this:
```
Comment (da-tanabe) [inline README.md:85]: This might be better as a permalink (instead of pointing to `master`).

File: README.md [hunk 4]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+message queues. Based on the specific event handlers that the
+integration registers, the integration framwork configures itself
+appropriately to make those events known to the integration while it
+is running.
+
+The `env` argument is the environment in which the integration is
+running. This contains named fields with all of the integration
+configuration parameters, as well as access to various other features
+of the integration framework. These include in-memory queuing and
+metadata for the DAML model associated with the integration. Because
```
Comment (da-tanabe) [inline README.md]: ~DAML~ Daml

File: README.md [hunk 5]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+from [`IntegrationEnvironment`](https://github.com/digital-asset/daml-dit-if/blob/master/daml_dit_if/api/__init__.py#L280),
+and if a specific subclass is not specified, the integration will
+receive an instance of `IntegrationEnvironment` as its environment
+when it starts up.  The syntax for `env_class` is the same as the
+syntax for `entrypoint` : `$PYTHON_PACKAGE_NAME:$FUNCTION_NAME`.
 
-def integration_ledger_event_log_main(
+## Logging
+
+DABL integrations use the default Python logging package, and the
```
Comment (da-tanabe) [inline README.md]: ~DABL~ Daml Hub

File: README.md [hunk 6]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+framework provides specific support for controlling log level at
```
Comment (da-tanabe) [inline README.md]: ~specific~

File: README.md [hunk 7]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+runtime. To integrate properly with this logic, it is important that
+integrations use the standard mechanism for accessing the integration
+logger. This logger is switched from `INFO` level to `DEBUG` level at
+a `DABL_LOG_LEVEL` setting of 10 or above.
```
Comment (da-tanabe) [inline README.md]: Heh…and what to do about `DABL_LOG_LEVEL`…

**Review by maxdeliso-da — APPROVED** [review]
lgtm

File: README.md [hunk 8]
```diff
@@ -1,59 +1,72 @@
...
+
+Integrations are packaged in [DIT files](https://github.com/digital-asset/daml-dit-api),
+built using the [`ddit` build tool](https://github.com/digital-asset/daml-dit-ddit),
+and can be deployed into Daml Hub using the same upload mechanism as other
+artifacts (Daml Models, bots, etc.). Daml Hub also has an 'arcade' facility
+that uses a public [GitHub repository](https://github.com/digital-asset/daml-dit-arcade-index)
+to maintain list of integrations and sample apps that can be deployed through
+a single click on the [console user interface](https://hub.daml.com/docs/quickstart).
+
+Logically speaking, DABL integrations package their Python
```
Comment (alexgraham-da) [inline README.md]: ~DABL~ Daml Hub

File: README.md [hunk 9]
```diff
@@ -1,59 +1,72 @@
...
+implementation code alongside metadata describing available
+integration types and any other resources required to make the
+integration operate. Integrations usually require Daml models to
+represent their data model, and often require external Python
+dependencies specified through a `requirements.txt` file. Both of
+these are examples of additional resourced that might be bundled into
```
Comment (alexgraham-da) [inline README.md]: ~resourced~ resources

File: README.md [hunk 10]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
 
+def integration_ledger_event_log_main(
+        env: 'IntegrationLedgerEventLogEnv',
+        events: 'IntegrationEvents'):
+
+```
+
+The `events` argument is an instance of [`IntegrationEvents`](https://github.com/digital-asset/daml-dit-if/blob/master/daml_dit_if/api/__init__.py#L272)
+containing a number of [function decorators](https://www.python.org/dev/peps/pep-0318/)
+that can be used to register event handlers. This represents the bulk
```
Comment (alexgraham-da) [inline README.md]: It's only a few paragraphs down, but possibly link "event handlers" to the `## Integration Event Handlers` section?

File: README.md [hunk 11]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+```
+
+## Integration Event Handlers
+
+Integrations are purely event driven and may only take action in
+response to an event notification from the framework. Integrations
+register their interest in given events by decorating custom functions
+with decorators provided to the integration when it is starting
+up. These decorators are found in the `IntegrationEvents` instance
+passed to te entrypoint function. As an example, the
```
Comment (alexgraham-da) [inline README.md]: ~te~ the

File: README.md [hunk 12]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+* **Ledger** - [DAZL](https://github.com/digital-asset/dazl-client) ledger events. (Contract Archived, Contract Created, Transaction Boundaries, etc.)
+* **Webhook** - Inbound HTTPS requests from the outside world. (GET or POST)
+* **Time** - Periodic timer events. (Useful to poll an external system, etc.)
+* **Queue** - In-memory message queue events. (Useful when none of the other types apply.)
+
+### Ledger Events
+
+The framework provides ledger event handlers for ledger initializtion
+events, transaction boundaries, and contract create and archived
+events.  These events are all subject to the the DAML ledger visiblity
```
Comment (alexgraham-da) [inline README.md]: ~DAML~ Daml

File: README.md [hunk 13]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+considered a minimum. The event handler for a given timer will not be
+called re-entrantly.
+
+### Queue Events
+
+The integration framework also provides in-memory queues and will call
+queue event handlers for message placed on those queues. The intent of
+this capability is to provide a way for integrations to respond to
+types of events that the other event handlers do not cover.  An
+example of this is the the Symphony intgration, which uses queue
```
Comment (alexgraham-da) [inline README.md]: ~the the~ the
~intgration~ integration

File: README.md [hunk 14]
```diff
@@ -82,132 +94,451 @@ integration_types:
...
+events to accept and process incoming messages received over a
+websocket. Because there is no websocket event hander type, the socket
```
Comment (alexgraham-da) [inline README.md]: This isn’t entirely accurate - the Symphony integration uses event listeners from Symphony’s Python API client library (which I believe uses polling). 

For the purposes of this doc however, the reasoning behind using the integration’s queue events are the same as if it was websockets directly. The Exberry integration does use websockets with the integration queue.

**Review by alexgraham-da — APPROVED** [review]
A few spelling comments and a clarification with the Symphony example, but otherwise this will be a very helpful reference!

File: README.md (hunk 2 above)
Comment (mschaef-da) [reply README.md]: Made this more clear.

File: README.md (hunk 3 above)
Comment (mschaef-da) [reply README.md:85]: I see your point, but I wrote it this way, preferring to link to the current latest.

File: README.md (hunk 6 above)
Comment (mschaef-da) [reply README.md]: Minus a lot of others too.

File: README.md (hunk 10 above)
Comment (mschaef-da) [reply README.md]: Great idea.

File: README.md (hunk 13 above)
Comment (mschaef-da) [reply README.md]: I've also hyperlinked to the Symphony repository.

File: README.md (hunk 14 above)
Comment (mschaef-da) [reply README.md]: Good catch. I've reworded. Thanks!

**Pull request merged**